server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/NetworkIsInWhitelist.java

```diff
@@ -80,61 +78,40 @@ public void init() throws MessagingException {
-
-            if (conn == null) {
-                conn = datasource.getConnection();
-            }
-
-            List<String> nets = new ArrayList<>();
-            try {
-                if (selectStmt == null) {
-                    selectStmt = conn.prepareStatement(selectNetworks);
-                }
-                selectStmt.setString(1, recipientUser);
-                selectStmt.setString(2, recipientHost);
-                selectRS = selectStmt.executeQuery();
-                while (selectRS.next()) {
-                    nets.add(selectRS.getString(1));
-                }
-            } finally {
-                jdbcUtil.closeJDBCResultSet(selectRS);
-                jdbcUtil.closeJDBCStatement(selectStmt);
-            }
-            NetMatcher matcher = new NetMatcher(nets, dns);
-            boolean matched = matcher.matchInetNetwork(mail.getRemoteAddr());
-
-            if (!matched) {
-                try {
-                    selectStmt = conn.prepareStatement(selectNetworks);
-                    selectStmt.setString(1, "*");
-                    selectStmt.setString(2, recipientHost);
-                    selectRS = selectStmt.executeQuery();
-                    nets = new ArrayList<>();
-                    while (selectRS.next()) {
-                        nets.add(selectRS.getString(1));
-                    }
-                } finally {
-                    jdbcUtil.closeJDBCResultSet(selectRS);
-                    jdbcUtil.closeJDBCStatement(selectStmt);
-                }
-                matcher = new NetMatcher(nets, dns);
-                matched = matcher.matchInetNetwork(mail.getRemoteAddr());
-            }
-            return matched;
+        String recipientUser = recipientMailAddress.getLocalPart().toLowerCase(Locale.US);
+        String recipientHost = recipientMailAddress.getDomain().asString();
+        try (Connection conn = datasource.getConnection()) {
+            return Flux
+                .concat(
+                    Mono.fromCallable(() -> loadNetworks(recipientUser, recipientHost, conn)),
+                    Mono.fromCallable(() -> loadNetworksForHost(recipientHost, conn)))
+                .map(nets -> new NetMatcher(nets, dns))
+                .any(matcher -> matcher.matchInetNetwork(mail.getRemoteAddr()))
+                .block();
         } catch (SQLException sqle) {
             LOGGER.error("Error accessing database", sqle);
             throw new MessagingException("Exception thrown", sqle);
-        } finally {
-            theJDBCUtil.closeJDBCConnection(conn);
         }
     }
 
+    private List<String> loadNetworksForHost(String recipientHost, Connection conn) throws SQLException {
+        return loadNetworks("*", recipientHost, conn);
+    }
+
+    private List<String> loadNetworks(String recipientUser, String recipientHost, Connection conn) throws SQLException {
+        List<String> nets = new ArrayList<>();
```

Comment:
Maybe `LinkedList` is better for add/remove than `ArrayList`

Reply:
There's no removal and appending is probably faster on an ArrayList.
BTW, I don't expect a lot of entries here so it should not make any difference in real life.